README.md
[![Build Status](https://travis-ci.org/platanus/nchan-prometheus-exporter.svg?branch=master)](https://travis-ci.org/platanus/nchan-prometheus-exporter) [![](https://images.microbadger.com/badges/version/platanus/nchan-prometheus-exporter.svg)](https:/hub.docker.com/r/platanus/nchan-prometheus-exporter) [![Go Report Card](https://goreportcard.com/badge/github.com/platanus/nchan-prometheus-exporter)](https://goreportcard.com/report/github.com/platanus/nchan-prometheus-exporter)

# Nchan Prometheus Exporter

Nchan Prometheus exporter makes it possible to monitor Nchan using Prometheus.

## Overview

[Nchan](http://nchan.io) is a scalable, flexible pub/sub server for the modern web, built as a module for the Nginx web server. It provides metrics via de [nchan_stub_status page](https://nchan.io/#nchan_stub_status-stats). Nchan Prometheus exporter fetches the metrics from a single Nchan, converts the metrics into appropriate Prometheus metrics types and finally exposes them via an HTTP server to be collected by [Prometheus](https://prometheus.io/).

## Getting Started

In this section, we show how to quickly run Nchan Prometheus Exporter for Nchan.

### Prerequisites

We assume that you have already installed Prometheus and Nchan. Additionally, you need to:
* Expose the built-in metrics in Nchan:
    * For NCHAN, expose the [nchan_stub_status page](https://nchan.io/#nchan_stub_status-stats) at `/nchan_stub_status` on port `8080`
    * (optional): For NGINX, expose the [stub_status page](http://nginx.org/en/docs/http/ngx_http_stub_status_module.html#stub_status) at `/nginx_stub_status` on port `8080`.
* Configure Prometheus to scrape metrics from the server with the exporter. Note that the default scrape port of the exporter is `9113` and the default metrics path -- `/metrics`.

### Running the Exporter in a Docker Container

To start the exporter we use the [docker run](https://docs.docker.com/engine/reference/run/) command.

* To export Nchan metrics, run:
    ```
    $ docker run -p 9113:9113 platanus/nchan-prometheus-exporter:0.1.0 -scrape-uri http://<nchan>:8080/nchan_stub_status
    ```

* (optional): To additionaly export Nginx metrics, run:
    ```
    $ docker run -p 9113:9113 platanus/nchan-prometheus-exporter:0.1.0 -scrape-uri http://<nchan>:8080/nchan_stub_status -nginx -nginx.scrape-uri http://<nchan>:8080/nginx_stub_status
    ```

> where `<nchan>` is the IP address/DNS name, through which Nchan is available.

### Running the Exporter Binary

* To export Nchan metrics, run:
    ```
    $ nchan-prometheus-exporter -nchan.scrape-uri http://<nchan>:8080/nchan_stub_status
    ```

* (optional): To additionaly export Nginx metrics, run:
    ```
    $ nchan-prometheus-exporter -nchan.scrape-uri http://<nchan>:8080/nchan_stub_status -nginx -nginx.scrape-uri http://<nchan>:8080/nginx_stub_status
    ```

> where `<nchan>` is the IP address/DNS name, through which Nchan is available.

**Note**. The `nchan-prometheus-exporter` is not a daemon. To run the exporter as a system service (daemon), configure the init system of your Linux server (such as systemd or Upstart) accordingly. Alternatively, you can run the exporter in a Docker container.

## Usage

### Command-line Arguments

```
Usage of ./nchan-prometheus-exporter:
  -scrape-uri string
        A URI for scraping Nchan metrics.
        The nchan_stub_status page must be available through the URI. The default value can be overwritten by SCRAPE_URI environment variable. (default "http://127.0.0.1:8080/nchan_stub_status")
  -nginx bool
        Start the exporter with NGINX metrics support. The default value can be overwritten by NGINX environment variable.
  -nginx.scrape-uri string
        A URI for scraping NGINX metrics.
        For NGINX, the stub_status page must be available through the URI. The default value can be overwritten by NGINX_SCRAPE_URI environment variable. (default "http://127.0.0.1:8080/nginx_stub_status")
  -ssl-verify
        Perform SSL certificate verification. The default value can be overwritten by SSL_VERIFY environment variable.
  -web.listen-address string
        An address to listen on for web interface and telemetry. The default value can be overwritten by LISTEN_ADDRESS environment variable. (default ":9113")
  -web.telemetry-path string
        A path under which to expose metrics. The default value can be overwritten by TELEMETRY_PATH environment variable. (default "/metrics")
```

### Exported Metrics

* For Nchan, all nchan_stub_status metrics are exported. Connect to the `/metrics` page of the running exporter to see the complete list of metrics along with their descriptions.

* For NGINX, all stub_status metrics are exported. Connect to the `/metrics` page of the running exporter to see the complete list of metrics along with their descriptions.

### Troubleshooting

The exporter logs errors to the standard output. When using Docker, if the exporter doesn’t work as expected, check its logs using [docker logs](https://docs.docker.com/engine/reference/commandline/logs/) command.

## Releases

For each release, we publish the corresponding Docker image at `platanus/nchan-prometheus-exporter` [DockerHub repo](https://hub.docker.com/r/platanus/nchan-prometheus-exporter/) and the binaries on the GitHub [releases page](https://github.com/platanus/nchan-prometheus-exporter/releases).

## Building the Exporter

You can build the exporter using the provided Makefile. Before building the exporter, make sure the following software is installed on your machine:
* make
* git
* Docker for building the container image
* Go for building the binary

### Building the Docker Image

To build the Docker image with the exporter, run:
```
$ make container
```

Note: go is not required, as the exporter binary is built in a Docker container. See the [Dockerfile](Dockerfile).

### Building the Binary

To build the binary, run:
```
$ make
```

Note: the binary is built for the OS/arch of your machine. To build binaries for other platforms, see the [Makefile](Makefile).

The binary is built with the name `nchan-prometheus-exporter`.

## Credits

Thank you [contributors](https://github.com/platanus/nchan-prometheus-exporter/graphs/contributors)!

<img src="http://platan.us/gravatar_with_text.png" alt="Platanus" width="250"/>

Nchan Prometheus Exporter is maintained by [platanus](http://platan.us).

## License

Cordova Plugin Flavors is © 2017 platanus, spa. It is free software and may be redistributed under the terms specified in the LICENSE file.


CHANGELOG.md
# Changelog

## 0.2.0

- [1](https://github.com/platanus/nchan-prometheus-exporter/pull/1) Add support for shared memory limit metrics from nchan >= 0.1.10

## 0.1.0

* Initial release.

nchanClient/nchan_test.go
package nchanClient

import "testing"

const validStabStats = "total published messages: 123\nstored messages: 54353\nshared memory used: 12K\nshared memory limit: 131072K\nchannels: 34\nsubscribers: 5434535\nredis pending commands: 48\nredis connected servers: 65\ntotal interprocess alerts received: 43\ninterprocess alerts in transit: 654\ninterprocess queued alerts: 765\ntotal interprocess send delay: 534\ntotal interprocess receive delay: 46\nnchan version: 1.1.5\n"

func TestParseStubStatsValidInput(t *testing.T) {
	var tests = []struct {
		input          []byte
		expectedResult StubStats
		expectedError  bool
	}{
		{
			input: []byte(validStabStats),
			expectedResult: StubStats{
				Redis: StubRedis{
					PendingCommands:  48,
					ConnectedServers: 65,
				},
				Channels:          34,
				Subscribers:       5434535,
				SharedMemoryUsed:  12,
				SharedMemoryLimit: 131072,
				Interprocess: StubInterprocess{
					AlertsInTransit:     654,
					QueuedAlerts:        765,
					TotalAlertsReceived: 43,
					TotalSendDelay:      534,
					TotalReceiveDelay:   46,
				},
				Messages: StubMessages{
					TotalPublished: 123,
					Stored:         54353,
				},
			},
			expectedError: false,
		},
		{
			input:         []byte("invalid-stats"),
			expectedError: true,
		},
	}

	for _, test := range tests {
		var result StubStats

		err := parseStubStats(test.input, &result)

		if err != nil && !test.expectedError {
			t.Errorf("parseStubStats() returned error for valid input %q: %v", string(test.input), err)
		}

		if !test.expectedError && test.expectedResult != result {
			t.Errorf("parseStubStats() result %v != expected %v for input %q", result, test.expectedResult, test.input)
		}
	}
}


nginxClient/nginx_test.go
package nginxClient

import "testing"

const validStabStats = "Active connections: 1457 \nserver accepts handled requests\n 6717066 6717066 65844359 \nReading: 1 Writing: 8 Waiting: 1448 \n"

func TestParseStubStatsValidInput(t *testing.T) {
	var tests = []struct {
		input          []byte
		expectedResult StubStats
		expectedError  bool
	}{
		{
			input: []byte(validStabStats),
			expectedResult: StubStats{
				Connections: StubConnections{
					Active:   1457,
					Accepted: 6717066,
					Handled:  6717066,
					Reading:  1,
					Writing:  8,
					Waiting:  1448,
				},
				Requests: 65844359,
			},
			expectedError: false,
		},
		{
			input:         []byte("invalid-stats"),
			expectedError: true,
		},
	}

	for _, test := range tests {
		var result StubStats

		err := parseStubStats(test.input, &result)

		if err != nil && !test.expectedError {
			t.Errorf("parseStubStats() returned error for valid input %q: %v", string(test.input), err)
		}

		if !test.expectedError && test.expectedResult != result {
			t.Errorf("parseStubStats() result %v != expected %v for input %q", result, test.expectedResult, test.input)
		}
	}
}


nchanClient/nchan.go
package nchanClient

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"strconv"
	"strings"
)

// NchanClient allows you to fetch Nchan metrics from the stub_status page.
type NchanClient struct {
	apiEndpoint string
	httpClient  *http.Client
}

// StubStats represents Nchan stub_status metrics.
type StubStats struct {
	Redis             StubRedis
	Interprocess      StubInterprocess
	Messages          StubMessages
	Channels          int64
	Subscribers       int64
	SharedMemoryUsed  int64
	SharedMemoryLimit int64
}

// StubRedis represents redis related metrics.
type StubRedis struct {
	PendingCommands  int64
	ConnectedServers int64
}

// StubInterprocess represents interprocess related metrics.
type StubInterprocess struct {
	AlertsInTransit     int64
	QueuedAlerts        int64
	TotalAlertsReceived int64
	TotalSendDelay      int64
	TotalReceiveDelay   int64
}

// StubMessages represents messages related metrics.
type StubMessages struct {
	TotalPublished int64
	Stored         int64
}

// NewNchanClient creates an NchanClient.
func NewNchanClient(httpClient *http.Client, apiEndpoint string) (*NchanClient, error) {
	client := &NchanClient{
		apiEndpoint: apiEndpoint,
		httpClient:  httpClient,
	}

	if _, err := client.GetStubStats(); err != nil {
		return nil, fmt.Errorf("Failed to create NchanClient: %v", err)
	}

	return client, nil
}

// GetStubStats fetches the stub_status metrics.
func (client *NchanClient) GetStubStats() (*StubStats, error) {
	resp, err := client.httpClient.Get(client.apiEndpoint)
	if err != nil {
		return nil, fmt.Errorf("failed to get %v: %v", client.apiEndpoint, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("expected %v response, got %v", http.StatusOK, resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read the response body: %v", err)
	}

	var stats StubStats
	err = parseStubStats(body, &stats)
	if err != nil {
		return nil, fmt.Errorf("failed to parse response body %q: %v", string(body), err)
	}

	return &stats, nil
}

func parseStubStats(data []byte, stats *StubStats) error {
	dataStr := string(data)

	parts := strings.Split(dataStr, "\n")
	if len(parts) != 15 {
		return fmt.Errorf("invalid input %q", dataStr)
	}

	totalPublishedMessagesParts := strings.Split(strings.TrimSpace(parts[0]), " ")
	if len(totalPublishedMessagesParts) != 4 {
		return fmt.Errorf("invalid input for total published messages %q", parts[0])
	}

	totalPublishedMessages, err := strconv.ParseInt(totalPublishedMessagesParts[3], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for total published messages %q: %v", totalPublishedMessagesParts[3], err)
	}
	stats.Messages.TotalPublished = totalPublishedMessages

	storedMessagesParts := strings.Split(strings.TrimSpace(parts[1]), " ")
	if len(storedMessagesParts) != 3 {
		return fmt.Errorf("invalid input for stored messages %q", parts[1])
	}

	storedMessages, err := strconv.ParseInt(storedMessagesParts[2], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for stored messages %q: %v", storedMessagesParts[2], err)
	}
	stats.Messages.Stored = storedMessages

	sharedMemoryUsedParts := strings.Split(strings.TrimSpace(parts[2]), " ")
	if len(sharedMemoryUsedParts) != 4 {
		return fmt.Errorf("invalid input for shared memory used %q", parts[2])
	}

	sharedMemoryUsed, err := strconv.ParseInt(strings.Replace(sharedMemoryUsedParts[3], "K", "", 1), 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for shared memory used %q: %v", sharedMemoryUsedParts[3], err)
	}
	stats.SharedMemoryUsed = sharedMemoryUsed

	sharedMemoryLimitParts := strings.Split(strings.TrimSpace(parts[3]), " ")
	if len(sharedMemoryLimitParts) != 4 {
		return fmt.Errorf("invalid input for shared memory limit %q", parts[3])
	}

	sharedMemoryLimit, err := strconv.ParseInt(strings.Replace(sharedMemoryLimitParts[3], "K", "", 1), 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for shared memory limit %q: %v", sharedMemoryLimitParts[3], err)
	}
	stats.SharedMemoryLimit = sharedMemoryLimit

	channelsParts := strings.Split(strings.TrimSpace(parts[4]), " ")
	if len(channelsParts) != 2 {
		return fmt.Errorf("invalid input for channels %q", parts[4])
	}

	channels, err := strconv.ParseInt(channelsParts[1], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for channels %q: %v", channelsParts[1], err)
	}
	stats.Channels = channels

	subscribersParts := strings.Split(strings.TrimSpace(parts[5]), " ")
	if len(subscribersParts) != 2 {
		return fmt.Errorf("invalid input for subscribers %q", parts[5])
	}

	subscribers, err := strconv.ParseInt(subscribersParts[1], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for subscribers %q: %v", subscribersParts[1], err)
	}
	stats.Subscribers = subscribers

	redisPendingCommandsParts := strings.Split(strings.TrimSpace(parts[6]), " ")
	if len(redisPendingCommandsParts) != 4 {
		return fmt.Errorf("invalid input for redis pending commands %q", parts[6])
	}

	redisPendingCommands, err := strconv.ParseInt(redisPendingCommandsParts[3], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for redis pending commands %q: %v", redisPendingCommandsParts[3], err)
	}
	stats.Redis.PendingCommands = redisPendingCommands

	redisConnectedServersParts := strings.Split(strings.TrimSpace(parts[7]), " ")
	if len(redisConnectedServersParts) != 4 {
		return fmt.Errorf("invalid input for redis connected servers %q", parts[7])
	}

	redisConnectedServers, err := strconv.ParseInt(redisConnectedServersParts[3], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for redis connected servers %q: %v", redisConnectedServersParts[3], err)
	}
	stats.Redis.ConnectedServers = redisConnectedServers

	totalInterprocessAlertsReceivedParts := strings.Split(strings.TrimSpace(parts[8]), " ")
	if len(totalInterprocessAlertsReceivedParts) != 5 {
		return fmt.Errorf("invalid input for interprocess total alerts received %q", parts[8])
	}

	totalInterprocessAlertsReceived, err := strconv.ParseInt(totalInterprocessAlertsReceivedParts[4], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for interprocess total alerts received %q: %v", totalInterprocessAlertsReceivedParts[4], err)
	}
	stats.Interprocess.TotalAlertsReceived = totalInterprocessAlertsReceived

	interprocessAlertsInTransitParts := strings.Split(strings.TrimSpace(parts[9]), " ")
	if len(interprocessAlertsInTransitParts) != 5 {
		return fmt.Errorf("invalid input for interprocess alerts in transit %q", parts[9])
	}

	interprocessAlertsInTransit, err := strconv.ParseInt(interprocessAlertsInTransitParts[4], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for interprocess alerts in transit %q: %v", interprocessAlertsInTransitParts[4], err)
	}
	stats.Interprocess.AlertsInTransit = interprocessAlertsInTransit

	interprocessQueuedAlertsParts := strings.Split(strings.TrimSpace(parts[10]), " ")
	if len(interprocessQueuedAlertsParts) != 4 {
		return fmt.Errorf("invalid input for interprocess queued alerts %q", parts[10])
	}

	interprocessQueuedAlerts, err := strconv.ParseInt(interprocessQueuedAlertsParts[3], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for interprocess queued alerts %q: %v", interprocessQueuedAlertsParts[3], err)
	}
	stats.Interprocess.QueuedAlerts = interprocessQueuedAlerts

	totalInterprocessSendDelayParts := strings.Split(strings.TrimSpace(parts[11]), " ")
	if len(totalInterprocessSendDelayParts) != 5 {
		return fmt.Errorf("invalid input for interprocess total send delay %q", parts[11])
	}

	totalInterprocessSendDelay, err := strconv.ParseInt(totalInterprocessSendDelayParts[4], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for interprocess total send delay %q: %v", totalInterprocessSendDelayParts[4], err)
	}
	stats.Interprocess.TotalSendDelay = totalInterprocessSendDelay

	totalInterprocessReceiveDelayParts := strings.Split(strings.TrimSpace(parts[12]), " ")
	if len(totalInterprocessReceiveDelayParts) != 5 {
		return fmt.Errorf("invalid input for interprocess total receive delay %q", parts[12])
	}

	totalInterprocessReceiveDelay, err := strconv.ParseInt(totalInterprocessReceiveDelayParts[4], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for interprocess total receive delay %q: %v", totalInterprocessReceiveDelayParts[4], err)
	}
	stats.Interprocess.TotalReceiveDelay = totalInterprocessReceiveDelay

	return nil
}


nginxClient/nginx.go
package nginxClient

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"strconv"
	"strings"
)

// NginxClient allows you to fetch NGINX metrics from the stub_status page.
type NginxClient struct {
	apiEndpoint string
	httpClient  *http.Client
}

// StubStats represents NGINX stub_status metrics.
type StubStats struct {
	Connections StubConnections
	Requests    int64
}

// StubConnections represents connections related metrics.
type StubConnections struct {
	Active   int64
	Accepted int64
	Handled  int64
	Reading  int64
	Writing  int64
	Waiting  int64
}

// NewNginxClient creates an NginxClient.
func NewNginxClient(httpClient *http.Client, apiEndpoint string) (*NginxClient, error) {
	client := &NginxClient{
		apiEndpoint: apiEndpoint,
		httpClient:  httpClient,
	}

	if _, err := client.GetStubStats(); err != nil {
		return nil, fmt.Errorf("Failed to create NginxClient: %v", err)
	}

	return client, nil
}

// GetStubStats fetches the stub_status metrics.
func (client *NginxClient) GetStubStats() (*StubStats, error) {
	resp, err := client.httpClient.Get(client.apiEndpoint)
	if err != nil {
		return nil, fmt.Errorf("failed to get %v: %v", client.apiEndpoint, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("expected %v response, got %v", http.StatusOK, resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read the response body: %v", err)
	}

	var stats StubStats
	err = parseStubStats(body, &stats)
	if err != nil {
		return nil, fmt.Errorf("failed to parse response body %q: %v", string(body), err)
	}

	return &stats, nil
}

func parseStubStats(data []byte, stats *StubStats) error {
	dataStr := string(data)

	parts := strings.Split(dataStr, "\n")
	if len(parts) != 5 {
		return fmt.Errorf("invalid input %q", dataStr)
	}

	activeConsParts := strings.Split(strings.TrimSpace(parts[0]), " ")
	if len(activeConsParts) != 3 {
		return fmt.Errorf("invalid input for active connections %q", parts[0])
	}

	actCons, err := strconv.ParseInt(activeConsParts[2], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for active connections %q: %v", activeConsParts[2], err)
	}
	stats.Connections.Active = actCons

	miscParts := strings.Split(strings.TrimSpace(parts[2]), " ")
	if len(miscParts) != 3 {
		return fmt.Errorf("invalid input for connections and requests %q", parts[2])
	}

	acceptedCons, err := strconv.ParseInt(miscParts[0], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for accepted connections %q: %v", miscParts[0], err)
	}
	stats.Connections.Accepted = acceptedCons

	handledCons, err := strconv.ParseInt(miscParts[1], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for handled connections %q: %v", miscParts[1], err)
	}
	stats.Connections.Handled = handledCons

	requests, err := strconv.ParseInt(miscParts[2], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for requests %q: %v", miscParts[2], err)
	}
	stats.Requests = requests

	consParts := strings.Split(strings.TrimSpace(parts[3]), " ")
	if len(consParts) != 6 {
		return fmt.Errorf("invalid input for connections %q", parts[3])
	}

	readingCons, err := strconv.ParseInt(consParts[1], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for reading connections %q: %v", consParts[1], err)
	}
	stats.Connections.Reading = readingCons

	writingCons, err := strconv.ParseInt(consParts[3], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for writing connections %q: %v", consParts[3], err)
	}
	stats.Connections.Writing = writingCons

	waitingCons, err := strconv.ParseInt(consParts[5], 10, 64)
	if err != nil {
		return fmt.Errorf("invalid input for waiting connections %q: %v", consParts[5], err)
	}
	stats.Connections.Waiting = waitingCons

	return nil
}


exporter.go
package main

import (
	"crypto/tls"
	"flag"
	"log"
	"net/http"
	"os"
	"strconv"

	"github.com/platanus/nchan-prometheus-exporter/collector"
	"github.com/platanus/nchan-prometheus-exporter/nchanClient"
	"github.com/platanus/nchan-prometheus-exporter/nginxClient"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

func getEnv(key, defaultValue string) string {
	value, ok := os.LookupEnv(key)
	if !ok {
		return defaultValue
	}
	return value
}

func getEnvBool(key string, defaultValue bool) bool {
	value, ok := os.LookupEnv(key)
	if !ok {
		return defaultValue
	}
	b, err := strconv.ParseBool(value)
	if err != nil {
		log.Fatalf("Environment Variable value for %s must be a boolean", key)
	}
	return b
}

var (
	// Set during go build
	version   string
	gitCommit string

	// Defaults values
	defaultListenAddress  = getEnv("LISTEN_ADDRESS", ":9113")
	defaultMetricsPath    = getEnv("TELEMETRY_PATH", "/metrics")
	defaultScrapeURI      = getEnv("CRAPE_URI", "http://127.0.0.1:8080/nchan_stub_status")
	defaultNginx          = getEnvBool("NGINX", false)
	defaultNginxScrapeURI = getEnv("NGINX_SCRAPE_URI", "http://127.0.0.1:8080/stub_status")
	defaultSslVerify      = getEnvBool("SSL_VERIFY", true)

	// Command-line flags
	listenAddr = flag.String("web.listen-address", defaultListenAddress,
		"An address to listen on for web interface and telemetry. The default value can be overwritten by LISTEN_ADDRESS environment variable.")
	metricsPath = flag.String("web.telemetry-path", defaultMetricsPath,
		"A path under which to expose metrics. The default value can be overwritten by TELEMETRY_PATH environment variable.")
	scrapeURI = flag.String("scrape-uri", defaultScrapeURI,
		`A URI for scraping Nchan metrics.
	The nchan_stub_status page must be available through the URI. The default value can be overwritten by SCRAPE_URI environment variable.`)
	nginx = flag.Bool("nginx", defaultNginx,
		"Start the exporter with NGINX metrics support. The default value can be overwritten by NGINX environment variable.")
	nginxScrapeURI = flag.String("nginx.scrape-uri", defaultScrapeURI,
		`A URI for scraping NGINX metrics.
	The stub_status page must be available through the URI. The default value can be overwritten by NGINX_SCRAPE_URI environment variable.`)
	sslVerify = flag.Bool("ssl-verify", defaultSslVerify,
		"Perform SSL certificate verification. The default value can be overwritten by SSL_VERIFY environment variable.")
)

func main() {
	flag.Parse()

	log.Printf("Starting Nchan Prometheus Exporter Version=%v GitCommit=%v", version, gitCommit)

	registry := prometheus.NewRegistry()

	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: !*sslVerify},
	}

	nchanClient, err := nchanClient.NewNchanClient(&http.Client{Transport: tr}, *scrapeURI)
	if err != nil {
		log.Fatalf("Could not create Nchan Client: %v", err)
	}

	registry.MustRegister(collector.NewNchanCollector(nchanClient, "nchan"))

	if *nginx {
		nginxClient, err := nginxClient.NewNginxClient(&http.Client{Transport: tr}, *nginxScrapeURI)
		if err != nil {
			log.Fatalf("Could not create Nginx Client: %v", err)
		}

		registry.MustRegister(collector.NewNginxCollector(nginxClient, "nginx"))
	}

	http.Handle(*metricsPath, promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
			<head><title>Nchan Exporter</title></head>
			<body>
			<h1>Nchan Exporter</h1>
			<p><a href='/metrics'>Metrics</a></p>
			</body>
			</html>`))
	})
	log.Fatal(http.ListenAndServe(*listenAddr, nil))
}
